blockchain/chain/src/store/chain_store.rs

```diff
@@ -646,6 +663,102 @@ where
+        Apply(TipsetJsonRef<'a>),
+        Revert(TipsetJsonRef<'a>),
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct IndexToHeadChangeJson(pub usize, pub HeadChange);
+    impl<'a> From<&'a HeadChange> for HeadChangeJson<'a> {
+        fn from(wrapper: &'a HeadChange) -> Self {
+            match wrapper {
+                HeadChange::Current(tipset) => HeadChangeJson::Current(TipsetJsonRef(&tipset)),
+                HeadChange::Apply(tipset) => HeadChangeJson::Apply(TipsetJsonRef(&tipset)),
+                HeadChange::Revert(tipset) => HeadChangeJson::Revert(TipsetJsonRef(&tipset)),
+            }
+        }
+    }
+
+    #[derive(Clone)]
+    pub enum EventsPayload {
+        TaskCancel(usize, ()),
+        SubHeadChanges(IndexToHeadChangeJson),
+    }
+
+    impl EventsPayload {
+        pub fn sub_head_changes(&self) -> Option<&IndexToHeadChangeJson> {
+            match self {
+                EventsPayload::SubHeadChanges(s) => Some(s),
+                _ => None,
+            }
+        }
+
+        pub fn task_cancel(&self) -> Option<(usize, ())> {
+            match self {
+                EventsPayload::TaskCancel(val, _) => Some((*val, ())),
+                _ => None,
+            }
+        }
+    }
+
+    pub async fn sub_head_changes(
+        mut subscribed_head_change: Subscriber<HeadChange>,
+        heaviest_tipset: &Option<Arc<Tipset>>,
+        current_index: usize,
+        events_pubsub: Arc<RwLock<Publisher<EventsPayload>>>,
+    ) -> Result<usize, Error> {
+        let head = heaviest_tipset
+            .as_ref()
+            .ok_or_else(|| Error::Other("Could not get heaviest tipset".to_string()))?;
+
+        (*events_pubsub.write().await)
+            .publish(EventsPayload::SubHeadChanges(IndexToHeadChangeJson(
+                current_index,
+                HeadChange::Current(head.clone()),
+            )))
+            .await;
+        let subhead_sender = events_pubsub.clone();
+        let handle = task::spawn(async move {
+            while let Some(change) = subscribed_head_change.next().await {
+                let index_to_head_change = IndexToHeadChangeJson(current_index, change);
+                subhead_sender
+                    .write()
+                    .await
+                    .publish(EventsPayload::SubHeadChanges(index_to_head_change))
+                    .await;
+            }
+        });
+        let cancel_sender = events_pubsub.write().await.subscribe();
+        task::spawn(async move {
+            if let Some(EventsPayload::TaskCancel(_, ())) = cancel_sender
+                .filter(|s| {
+                    future::ready(
+                        s.task_cancel()
+                            .map(|s| s.0 == current_index)
+                            .unwrap_or_default(),
+                    )
+                })
+                .next()
+                .await
+            {
+                handle.cancel().await;
+            }
+        });
```

Comment:
why is a cancel event published on a pubsub type channel? When would you ever want to send something to cancel all subscriptions? I'm really confused on this pattern, or how it would be used, if you don't mind elaborating the intention

Reply:
The cancel even t is not sent to a publish. The channel that is done with the sub_head_changes is sent to this function with the appropriate id and the sub_head_changes method that matches up with this ID is cancelled. They can be a of sub_head_changes running at the same time. So this doesn't publish but checks if an event with the subscription ID has been passed through